# GeNA
`GeNA` (Genotype-Neighborhood Associations) is a method for identifying genetic variant associations to the abundance of cell states in single-cell datasets (cell state abundance quantitative trait loci, csaQTLs). Instead of testing associations to predefined cell types, `GeNA` identifies the granular cell states whose abundance is most associated with genetic variants. The scripts required to run `GeNA` are stored in this repo, with accompanying tutorials and user support information.

We have evaluated GeNA in simulation to assess calibration and statistical power and we have applied `GeNA` in a genome-wide survey to sc-mRNAseq profiling from a population cohort of 969 individuals. Scripts documenting our use of `GeNA` in these analyses for our manuscript are found in a separate repository, [immunogenomics/GeNA-applied](https://github.com/immunogenomics/GeNA-applied/). For more information about GeNA please refer to our preprint linked [here].

# Installation
To use `GeNA`, you can clone this repository and add it to your `PYTHONPATH`

# Tutorial
We illustrate how to use `GeNA` on a toy dataset of single-cell profiling and simulated genotypes here.

# Citation
If you use `GeNA` in your work, you can cite our preprint as: <>

# Contact
If you have questions about `GeNA` or require user support, please contact Laurie Rumker (Laurie_Rumker@hms.harvard.edu) or post an issue on this repo.
